components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -2378,6 +2416,206 @@ private static JWT encryptWithRSA(JWTClaimsSet jwtClaimsSet, JWEAlgorithm encryp
+            } catch (IdentityOAuth2Exception e) {
+                log.error(String.format("Failed to set x5c thumbprint as kid value due to %s", e));
+                kid = null;
+            }
+        }
+        return kid;
+    }
+
+    /**
+     * Get encryption jwk from JWKS list when JWKS Uri is given.
+     *
+     * @param jwksUri - JWKS Uri
+     * @return - encryption JWK from the jwks url
+     * @throws IdentityOAuth2Exception - IdentityOAuth2Exception
+     */
+    private static JWK getEncryptionJWKFromJWKS(String jwksUri) throws IdentityOAuth2Exception {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Attempting to retrieve encryption jwk from the Jwks uri: %s.", jwksUri));
+        }
+        try {
+            JWKSet publicKeys = JWKSet.load(new URL(jwksUri));
+            JWK jwk = null;
+            //Get the first encryption JWK from the list
+            List<JWK> jwkList = publicKeys.getKeys();
+
+            for (JWK currentJwk : jwkList) {
+                if (KeyUse.ENCRYPTION == currentJwk.getKeyUse()) {
+                    jwk = currentJwk;
+                    break;
+                }
+            }
+            // if no jwk found for encryption, check for jwk with sig
+            if (jwk == null) {
+                for (JWK currentJwk : jwkList) {
+                    if (KeyUse.SIGNATURE == currentJwk.getKeyUse()) {
+                        jwk = currentJwk;
+                        break;
+                    }
+                }
+            }
+            if (jwk == null) {
+                throw new IdentityOAuth2Exception(String.format("Failed to retrieve valid jwk from " +
+                        "jwks uri: %s", jwksUri));
+            }
+            return jwk;
+        } catch (ParseException | IOException e) {
+            throw new IdentityOAuth2Exception(String.format("Failed to retrieve jwk from " +
+                    "jwks uri: %s", jwksUri), e);
+        }
+    }
+
+    /**
+     * Get public certificate from JWK
+     *
+     * @param jwk
+     * @return
+     * @throws IdentityOAuth2Exception
+     */
+    private static X509Certificate getPublicCertFromJWK(JWK jwk) throws IdentityOAuth2Exception {
+
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Attempting to retrieve public certificate from the Jwk kid: %s.", jwk.getKeyID()));
+        }
+        X509Certificate certificate;
+        if (jwk != null) {
+            if (jwk.getParsedX509CertChain() != null) {
+                certificate = jwk.getParsedX509CertChain().get(0);
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("Retrieved the public signing certificate successfully from the " +
+                            "jwk : %s", jwk));
+                }
+                return certificate;
+            } else {
+                throw new IdentityOAuth2Exception("Failed to retrieve public certificate from jwk");
+            }
+        } else {
```

Comment:
can we simplify this if else  part? 
if(jwk == null){
throw new IdentityOAuth2Exception
}

// have implementation for jwk!=null part.

Reply:
simplified the if condition in f20fb01